File: ui/src/app/map/domain/map-click.ts
import {MapState} from "./map-state";
import {SelectedFeatureHolder} from "./selected-feature-holder";

import Select from 'ol/interaction/Select';
import {click} from 'ol/events/condition';
import Style from 'ol/style/Style';
import {SelectedFeature} from "./selected-feature";

export class MapClick {

  readonly interaction = new Select({
    condition: click,
    multi: false,
    style: new Style() // this overrides the normal openlayers default edit style
  });

  constructor(
    state: MapState,
    selectionHolder: SelectedFeatureHolder,
    refreshCallback: () => void) {

    this.interaction.on("select", (e) => this.selectListener(state, selectionHolder, refreshCallback, e));
  }

  private selectListener(
    mapState: MapState,
    selectionHolder: SelectedFeatureHolder,
    refreshCallback: () => void,
    e /*: ol.interaction.select.Event*/): boolean {
    for (let feature of e.deselected) {
      const layer = feature.get("layer");
      if (layer.endsWith("route")) {
        mapState.selectedRouteId = null;
      }
      else if (layer.endsWith("node")) {
        mapState.selectedNodeId = null;
      }
    }

    if (e.selected.isEmpty) {
      selectionHolder.select(null)
    }
    else {
      for (let feature of e.selected) {
        const layer = feature.get("layer");
        if (layer.endsWith("route")) {
          const id = feature.get("id");
          const id2 = id.substring(0, id.indexOf('-'));
          const name = feature.get("name");
          mapState.selectedRouteId = id2 + "-";
          const selection = new SelectedFeature(+id2, name, "route");
          selectionHolder.select(selection)
        }
        else if (layer.endsWith("node")) {
          mapState.selectedRouteId = null;
          const id = feature.get("id");
          mapState.selectedNodeId = id;
          const name = feature.get("name");
          const selection = new SelectedFeature(+id, name, "node");
          selectionHolder.select(selection);
        }
      }
    }

    refreshCallback();

    return true;
  }

}

File: production/client/src/main/scala/kpn/client/common/map/vector/MapClick.scala
// TODO migrate to Angular
package kpn.client.common.map.vector

class MapClick(state: MapState, selectionHolder: SelectedFeatureHolder, refreshCallback: () => Unit) {

  val interaction: ol.interaction.Select = {
    val select = new ol.interaction.Select(
      olx.interaction.SelectOptions(
        condition = ol.events.condition.Condition.click _,
        multi = false,
        style = ol.style.Style() // this overrides the normal openlayers default edit style
      )
    )
    select.on("select", selectListener _)
    select
  }

  private def selectListener(e: ol.interaction.select.Event): Boolean = {

    e.deselected.foreach { feature: ol.Feature =>
      val layer = feature.asInstanceOf[ol.render.Feature].get("layer").toString
      if (layer.endsWith("route")) {
        state.selectedRouteId = None
      }
      else if (layer.endsWith("node")) {
        state.selectedNodeId = None
      }
    }

    if (e.selected.isEmpty) {
      selectionHolder.select(None)
    }
    else {
      e.selected.foreach { feature: ol.Feature =>
        val layer = feature.asInstanceOf[ol.render.Feature].get("layer").toString
        if (layer.endsWith("route")) {
          val id = feature.asInstanceOf[ol.render.Feature].get("id").toString
          val id2 = id.takeWhile(_ != '-')
          val name = feature.asInstanceOf[ol.render.Feature].get("name").toString
          state.selectedRouteId = Some(id2 + "-")
          val selection = SelectedRoute(id2.toLong, name)
          selectionHolder.select(Some(selection))
        }
        else if (layer.endsWith("node")) {
          state.selectedRouteId = None
          val id = feature.asInstanceOf[ol.render.Feature].get("id").toString
          state.selectedNodeId = Some(id.toString)
          val name = feature.asInstanceOf[ol.render.Feature].get("name").toString
          val selection = SelectedNode(id.toLong, name)
          selectionHolder.select(Some(selection))
        }
      }
    }

    refreshCallback()

    true
  }

}
